vm-image-builder: Automate creating customized Kubevirt VM's images

ormergi opened the pull request:
This PR add tool for automating the process of creating new container-disk image for KubeVirt VMs
based on  http://github.com/ormergi/vm-image-builder

It is targeted for developers or anyone who would like to create customized ephemeral container disk image for Kubevirt VM's
As were presented at kubevirt/kubevirt#3834

This tool should be used on the user host, integrating this to CI will be presented in following PR.

This PR:
Fixes kubevirt/kubevirt#3903
Is needed to continue with kubevirt/kubevirt#3834 and  kubevirt/kubevirt#3850

Release note:
```
Automate creating customized KubeVirt container disk images.
```

ormergi commented:
/cc @EdDev

EdDev commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,59 @@
+  virsh destroy "$DOMAIN_NAME" || true
+  virsh undefine "$DOMAIN_NAME" || true
+
+  if [ $? -ne 0 ]; then
+    rm -f "$NEW_IMAGE"
+  fi
```
This is a common bash pattern:
```
[ $? -ne 0 ] && rm -f "$NEW_IMAGE"
```

But this will always run because you do `|| true`

ormergi replied:
so should I move it to line 6?

EdDev replied:
Just run it without any condition, the `-f` will always cause it to succeed.

ormergi replied:
I want to delete  NEW_IMAGE only if there was an error during the script invocation
this is why I added the condition in the first place, it should be moved to line 6
so `|| true` will not interfere with `$?`

EdDev commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,59 @@
+NEW_IMAGE=$2
+CLOUD_CONFIG_PATH=$3
+
+export OS_VARIANT=${OS_VARIANT:-fedora31}
```
Why 31 and not 32? Maybe even without a version? It is just a label, nothing more.

ormergi replied:
I will remove it in following commits

ormergi replied:
After few tests apparently this flag is necessary for cloud-init to work properly with virt-install
I will drop `export` and document this in the commit and readme

EdDev commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,59 @@
+readonly CLOUD_INIT_ISO="cloudinit.iso"
+
+if ! [ -f "${CLOUD_CONFIG_PATH}" ]; then
+  CLOUD_CONFIG_PATH="user-cloud-config"
```
I do not understand this. You use the same variable to provide the file instead of the provided path?

ormergi replied:
If the file path of `CLOUD_CONFIG_PATH` does not exist its sets `CLOUD_CONFIG_PATH` default file path.
There is an error here because the file "user-cloud-config" is not there I forgot to add it,
I should set `CLOUD_CONFIG_PATH` default to `example/example-cloud-config` file

EdDev commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,59 @@
+
+echo "Customize image by booting a VM with
+ the image and cloud-init disk
+ press ctrl+] to exit"
```
> press ctrl+] to exit

But you do not need this as you will not stay on the console, right?

ormergi replied:
You can run this script only, and you didnt add `sudo shutdown` in cloud-config file at `runCmd`
you will stay connected to the console until you close it manually.
This is good if you want to check what is the state of the image on live VM after cloud init provision.

EdDev commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,59 @@
+
+virt-sysprep -d $DOMAIN_NAME --operations machine-id,bash-history,logfiles,tmp-files,net-hostname,net-hwaddr,customize --hostname ""
+
+# Remove VM
+virsh destroy $DOMAIN_NAME || true
+virsh undefine $DOMAIN_NAME
+
+# Remove cloud-init image
+rm -rf $CLOUD_INIT_ISO
```
Won't this happen on exit anyway (based on the trap)?

ormergi replied:
Yes it will, at first I though it should be removed so it wont collide with `qemu-img convert`
but now I dont see how this could happen, this like should be removed.

EdDev commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,59 @@
+}
+
+IMAGE_PATH=$1
+NEW_IMAGE=$2
```
Is this also a path, right?

ormergi replied:
Yes this the path of the new qcow2 file that will be created

EdDev replied:
So if be consistent and add "PATH" to this name as well.
Maybe you should call them SRC and DST (instead of NEW) for name consistency.

ormergi replied:
How about `SOURCE_IMAGE` and `CUSTOMIZED_IMAGE`?

EdDev commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,36 @@
+  echo "[build-container-disk] cleanup"
+  rm -f Dockerfile
+
+  if [ $? -ne 0 ]; then
+    rm -f "$tar_file"
+    docker image rm "$full_image_tag"
+  fi
```
I do not understand this one. If the file removal done earlier fails, you do not remove the rest?
It also uses "-e" so it is expected to fail anyway on first error, I see no point to check for the returned value.

ormergi replied:
The idea is to  delete the exported image tar file and the image stored in local docker registry, if there is was error during  he script.
I will move lines 13-16 to line 10 so it will actually catch errors from the script and not from cleanup().

EdDev replied:
I'm not sure I understand what you just mentioned but please make sure to test your conditions to make sure they do what you indented them to do.

EdDev commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,36 @@
+}
+
+image_name=$1
+tag=$2
```
Why a tag is needed? Isn't there a default? (latest)

ormergi replied:
To enable users to pick their own tag

EdDev replied:
Lets continue the "why"... Why would the users pick their own tag?
I fail to understand the scenario.

ormergi replied:
You could create few container-disk images and test them instead of overriding the default `latest` tag, for example: 
creating container-disk with the name "fedora" and the tag will be "32"
then you could test this image by creating a VM and use it:
on VM yaml file:
```...
  image: localhost:<port>/fedora:32
...```
You could create another version of the image without loosing `fedora:32` image, or changing its tag manually
let say we created another version with the tag `fedora:tests`, you could use it in a VM like so:
```...
  image: localhost:<port>/fedora:tests
```

EdDev commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,36 @@
+tag=$2
+vm_image_file=$3
+
+export TEMPLATE_DOCKERFILE=${TEMPLATE_DOCKERFILE:-$SCRIPT_DIR/Dockerfile.containerdisk}
```
Why do you need to export this? Why is using it later?

ormergi replied:
You right it should not be exported, I will remove `export` in following commits

EdDev commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,36 @@
+
+trap 'cleanup' EXIT
+
+export IMAGE=${vm_image_file}
```
Why do you need to export this?

ormergi replied:
You right it should not be exported, I will remove export in following commits

ormergi replied:
Sorry I mislead you, `IMAGE` should be exported because we use `envsubst` to replace the placeholder with the image file path

EdDev commented on cluster-provision/images/vm-image-builder/test-containerdisk.sh:
```diff
@@ -0,0 +1,67 @@
+
+readonly KUBEVIRT_URL="https://github.com/kubevirt/kubevirt"
+
+trap 'cleanup' EXIT SIGINT
```
I do not think you need SIGINT, EXIT should cover exit reasons. (please confirm)

EdDev commented on cluster-provision/images/vm-image-builder/test-containerdisk.sh:
```diff
@@ -0,0 +1,67 @@
+
+function cleanup() {
+    echo "[test-containerdisk] cleanup"
+    if [ -z $DEBUG ]; then
```
"DEBUG" merits an explanation in the commit or here.

ormergi replied:
I will add it to the commit message
Basically DEBUG lets you keep the cluster created by this script, the VMI with the customized image and the customized image tar file,
So you could debug the image/VM if there was an error during the script invocation, for example:
 if the VMI that created with the new image wont be ready, you could debug what happen to the VMI and console to the VMI and check.

EdDev commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+set -exuo pipefail
+
+SCRIPT_DIR="$(
```
Not sure what you are doing here, but sub-shelling is wasteful and looks strange.
Maybe this will help:
Getting the current folder (of where the script exists): `EXEC_PATH=$(dirname "$(realpath "$0")")`
Getting one folder above: `PARENT_PATH="$(dirname $EXEC_PATH)"`
Getting a folder under the current one: `FOO_PATH="$PWD/foo"`

ormergi replied:
I will simplify it in following commits.

EdDev commented:
> vm-image-builder: add test example

This one fetched kubevirt and tests against it? That cannot happen under kubevirtci.
Why would it even depend in kubevirt or kubevirtci? It needs to depend on the minimum, e.g. run it using "virsh" and check it is working (maybe create an artifact and check it is there). There is no need to raise a cluster to confirm the image is good.

ormergi commented:
> I do not understand the first commit.
> The fact that you fork the original repo should not be expressed in the code itself, not on the commit.
> 
> I do not think we can have such a dependency or to maintain it here.
> I guess the original repo can be kept in sync manually (or with such scripts) until we will decide to extract this code to its own repo (at which time we can just move your original repo into kubevirt).

I will remove this commit

ormergi commented:
> vm-image-builder: publish script:
> 
> I think this script does not deserve existence. Whoever needs to do this can call docker with the correct commands, we do not need to wrap these two lines.
> 
> Or do I miss something else here?

I will remove it and add what it does to create-push script

ormergi commented:
@EdDev I addressed most of the changes requested in new commits and updated the commit messages

oshoval commented on cluster-provision/images/vm-image-builder/create-containerdisk.sh:
```diff
@@ -0,0 +1,60 @@
+readonly build_directory="${IMAGE_NAME}_build"
+readonly new_vm_image_name="provisioned-image.qcow2"
+
+trap 'cleanup' EXIT SIGINT
```
just Exit is needed since it will run all the time

another thing is that -e might wont catch errors in functions according what read and saw once,
i would suggest to inject an error for example if VM_IMAGE doesnt exists 
and see if the cleanup is triggered
if it doesnt 
the -e should be replaced to -E and then it should catch it

ormergi replied:
I will remove `SIGINT`

If VM_IMAGE is removed, since `-u` flag is up the script stops with `unbound variable` error
If `-u` is removed as well `cleanup` still invokes
I am not familiar with `-E` what it does?

oshoval replied:
dont remove the var, remove the file the var point to.
trap might not catch errors that happen in functions even there is -e
-E should fix it if it indeed happen when u inject an error

ormergi replied:
`VM_IMAGE` is path to downloaded image, if you set it to bad path it still works and cleanup is invoked
I saw in man page that `-E` works with trap that catch `ERR` interrupts, I dont think we need it in this script

oshoval replied:
question is if the first error that happened was in a function, 
can you just add please `false` in a function and make sure it still runs indeed?

ormergi replied:
I checked and it works

oshoval replied:
thx

oshoval commented on cluster-provision/images/vm-image-builder/fedora-extended/cloud-config:
```diff
@@ -0,0 +1,26 @@
+  - sudo hostnamectl set-hostname ""
+  - sudo hostnamectl set-hostname "" --transient
+  - sudo systemctl start qemu-guest-agent
+  - sudo shutdown -h now
```
nit: newline

ormergi replied:
Done

oshoval commented on cluster-provision/images/vm-image-builder/fedora-extended/image-url:
```diff
@@ -0,0 +1 @@
+https://download.fedoraproject.org/pub/fedora/linux/releases/32/Cloud/x86_64/images/Fedora-Cloud-Base-32-1.6.x86_64.qcow2
```
nit: newline

ormergi replied:
Done

oshoval commented on cluster-provision/images/vm-image-builder/example/image-url:
```diff
@@ -0,0 +1 @@
+https://download.fedoraproject.org/pub/fedora/linux/releases/32/Cloud/x86_64/images/Fedora-Cloud-Base-32-1.6.x86_64.qcow2
```
nit: newline

ormergi replied:
Done

oshoval commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,53 @@
+  --name $DOMAIN_NAME \
+  --disk "$SOURCE_IMAGE_PATH",device=disk \
+  --disk "$CLOUD_INIT_ISO",device=cdrom \
+  --os-type Linux \
```
some os have a more specific os-type package, so maybe worth to parameterize it

oshoval commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,53 @@
+readonly DOMAIN_NAME="provision-vm"
+readonly CLOUD_INIT_ISO="cloudinit.iso"
+
+trap 'cleanup' EXIT SIGINT
```
remove here plz SIGINT as well

ormergi replied:
Done

ormergi replied:
I checked it again and `SIGINT` is necessary to invoke `cleanup` if the user interupt the script with CTRL+C

oshoval replied:
if so, consider please SIGQUIT as well
like ctrl c but with core dump, we dont use it most of the time, but its a good practice i believe to have a full picture

oshoval commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,53 @@
+  fi
+
+  rm -rf "$CLOUD_INIT_ISO"
+  virsh destroy "$DOMAIN_NAME" || true
```
all over, please keep one format ie
"${VAR}" and not just double quote, since you did use with `"{` in some places and indeed its the cleanest

ormergi replied:
Done
All scripts aligned to the format `"${VAR}"`

oshoval commented on cluster-provision/images/vm-image-builder/create-containerdisk.sh:
```diff
@@ -0,0 +1,60 @@
+
+trap 'cleanup' EXIT SIGINT
+
+pushd "${SCRIPT_PATH}"
```
please use '(' instead of the pushd + popd

ormergi replied:
Using `( ` `)` cuses all command inside to run in a subshell and it is not necessay in this script
Also I prefer using `pushd` and `popd` because it is more verbose commands

oshoval replied:
i dont see any reason to not use a sub shell,
its more nature here to use '(' without the push pop, and the ident will be more natural, now its not natural imo.
lets see what other reviewers think please.

ormergi replied:
By verbose I meant to the commands name, so instead of `(` we use `pushd` it make it more readable in my opinion.
If you dont mind lets wait for other reviewers

oshoval replied:
First time i had this dilemma i also thought like that, but then was asked to use `(`
and got used to it, we are adaptive, and now its self explanatory, i believe we should stick to a common practice,
it will save us precious time.

ormergi replied:
BTW I tried it and since using `( )` creating subshell, it is necessary to catch `ERR` interupts with `trap` and adding `-E` flag in order to invoke `cleanup` on errors
Maybe this explains why you saw scripts with `-e` and `trap` that wont catch errors

oshoval replied:
thx, ok lets then leave it as you did, just i am not sure the ident there is right to use.

oshoval commented on cluster-provision/images/vm-image-builder/create-containerdisk.sh:
```diff
@@ -0,0 +1,60 @@
+    curl -L $VM_IMAGE_URL -o $VM_IMAGE
+  fi
+
+  mkdir "${build_directory}"
```
mkdir -p , in case its re entrant

oshoval commented on cluster-provision/images/vm-image-builder/create-containerdisk.sh:
```diff
@@ -0,0 +1,60 @@
+    curl -L $VM_IMAGE_URL -o $VM_IMAGE
+  fi
+
+  mkdir "${build_directory}"
```
-p , in case the previous run was fine

ormergi replied:
I did not added `-p` so if a user run it again it wont delete the qcow2 image that was already created
What you think?

oshoval replied:
but wont it recreate it on the same name ?
i would then fail fast if the folder exists, if the case is what is above, or just `rm -rf` the folder if the above answer is yes

ormergi replied:
It is like you described, if you already created an image next time you run this script you will fail with an error saying directory already exists

ormergi replied:
Sorry for the noise, at first the idea was to keep files if something went wrong but it had no use
so i changed to delete file that were created in each phase is there is an error

oshoval replied:
i saw before that it clean just if error occour, but what if no error ?
i think that even if you choose to fail the test , i would prefer please something like
[ -d dir ] && echo "dir exist" && exit 1
(dont remember the exact check in the first one)

ormergi replied:
If no error occurs and the `...build` directory exists it will report that the directory exists and delete it, then on the second run it will run normally

oshoval commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,37 @@
+function cleanup() {
+  if [ $? -ne 0 ]; then
+    rm -f "$tar_file"
+    docker image rm -f "$full_image_tag"
```
if image doesnt exists , it wont fail ?
if it can fail add || true please

ormergi replied:
Since `-f` is mentioned `echo $?` returns `0` so we good

oshoval commented on cluster-provision/images/vm-image-builder/README.md:
```diff
@@ -0,0 +1,170 @@
+REPOSITORY - image repository
+
+This script expects: 
+IAMGE_NAME - container disk image name 
```
typo

ormergi replied:
Fixed

oshoval commented on cluster-provision/images/vm-image-builder/README.md:
```diff
@@ -0,0 +1,170 @@
+
+# Generated image .tar file
+$ ls "example-fedora_build/example-fedora-tests.tar"
+
```
space

ormergi replied:
Done

oshoval commented on cluster-provision/images/vm-image-builder/README.md:
```diff
@@ -0,0 +1,170 @@
+
+What this script does:
+- Creates temp directory and copy the source
+  VM image and '
```
`and '` ?

ormergi replied:
Fixed
I also arranged this file and removed redundant explanations

oshoval commented on cluster-provision/images/vm-image-builder/README.md:
```diff
@@ -0,0 +1,170 @@
+
+# You could send / store the image .tar file: 
+$ ls example_build/example-tests.tar
+
```
space

ormergi replied:
Done

oshoval commented on cluster-provision/images/vm-image-builder/README.md:
```diff
@@ -0,0 +1,170 @@
+
+# Create kubernetes cluster with Kubvirt
+git clone https://github.com/kubevirt/kubevirt --depth=1 --branch=master --single-branch kubevirt
+$ pushd kubevirt
```
add ``` please

oshoval commented on cluster-provision/images/vm-image-builder/README.md:
````diff
@@ -0,0 +1,170 @@
+
+## How to use:
+```bash
+
````
space

ormergi replied:
Done

ormergi commented:
Changes:
- Resolved all nits, formats and typos at README file, also rearranged it and rephrased few explanations
- Changed the format of using variables in all scripts to `"${VAR}"`
- Renamed `fedora-extended` to `fedora-sriov-lane` to describe better the propose of the files
- Changed `sudo shutdown -h now` to `sudo shutdown` in `example/cloug-config` `fedora-sriov-lane/cloud-config` files so cloud-init will finish its operation with no errors.
Using `shutdown` w/o `-h now` will shutdown a host after one minute

oshoval commented on cluster-provision/images/vm-image-builder/OWNERS:
```diff
@@ -0,0 +1,6 @@
+filters:
+  ".*":
+    reviewers:
+      - ormergi
```
i think we should use something like network-reviewers
(on both lines, as other OWNER files are doing please)

EdDev replied:
This adds ormergi to the reviewers and approvers of this specific folder. I do not see why it is worth creating a special group for this at the moment, especially when only one is included in it.

oshoval replied:
no problem, just saying what i saw on other files in some kubevirt repos, and on other habits from the past.

EdDev reviewed (commented):
Commit "vm-image-builder: build script": typo, `Dacording`.

EdDev commented on cluster-provision/images/vm-image-builder/build-containerdisk.sh:
```diff
@@ -0,0 +1,37 @@
+
+trap 'cleanup' EXIT
+
+temp_dir=$(mktemp -d -p /tmp -t build.container-disk.XXXX)
```
Who cleans the temp directory at the end?

oshoval replied:
there is a trap but look on 
https://github.com/kubevirt/kubevirt/pull/4115#discussion_r484034330
bottom line we should return SIGTERM

EdDev replied:
ohh I missed line 12.
SIGTERM is discusses in a different thread, no need to mention it here.

EdDev reviewed (commented):
Commit "vm-image-builder: Automate VM customization flow": Typo, `containes`.

EdDev commented on cluster-provision/images/vm-image-builder/README.md:
```diff
@@ -0,0 +1,162 @@
+
+This script requires the following packages:
+- cloud-utils
+- docker-ce
```
You should consider changing to podman in a following change.

ormergi replied:
podman can be used as well I will add it to README

ormergi replied:
I will test it and consider a following PR

oshoval commented on cluster-provision/images/vm-image-builder/create-containerdisk.sh:
```diff
@@ -0,0 +1,60 @@
+readonly build_directory="${IMAGE_NAME}_build"
+readonly new_vm_image_name="provisioned-image.qcow2"
+
+trap 'cleanup' EXIT
```
you said you need SIGINT right ?
also please consider adding SIGTERM, see
https://github.com/kubevirt/kubevirt/pull/4115#discussion_r484034330

EdDev replied:
But we use `bash` here, not anything else.

oshoval replied:
wouldnt it might run on prow one day ?
better be safe and robust i think, in any case

EdDev replied:
It will probably run in a container and there is little chance we will not be fedora/centos base.

oshoval replied:
imho until it has one click to run in a container i would suggest to be robust,
the change which is needed is minor and harmless from my point of view

EdDev replied:
@oshoval , there is no robustness here, the script is limited to bash and it says so on the first line of the file.

ormergi replied:
No need for `SIGINT`/`SIGTERM` for this PR scripts, I checked and it works as expected with `EXIT`
See this [disscution](https://github.com/kubevirt/kubevirtci/pull/428#discussion_r476229164)

oshoval replied:
https://github.com/kubevirt/kubevirtci/pull/428#discussion_r476598843
you said the SIGINT is needed, if you double checked and its not needed, then remove but please update if possible so we know the behaviour.

SIGTERM is not needed because you have shebang of bash as Edy said.

ormergi replied:
Alright I will add it , it was two weeks ago and all those threads are confusing :confused:

oshoval commented on cluster-provision/images/vm-image-builder/customize-image.sh:
```diff
@@ -0,0 +1,53 @@
+readonly DOMAIN_NAME="provision-vm"
+readonly CLOUD_INIT_ISO="cloudinit.iso"
+
+trap 'cleanup' EXIT
```
there are multi traps, please see the discussion about SIGTERM and SIGINT 
(we will continue discussion on the other threads)

(wont write it on each one, please consider on each trap occurence)

ormergi commented:
Changes:
- fixed typos in commit messages
- added `SIGINT` where there is `trap` usage

EdDev commented:
> Changes:
> 
>     * fixed typos in commit messages
> 
>     * added `SIGINT` where there is `trap` usage

You have both rebased and added the changes, making it very hard now to see what changed.
Please do not mix both when pushing.

EdDev reviewed (approved):
Thank you, looks great.

ormergi commented:
> > Changes:
> > ```
> > * fixed typos in commit messages
> > 
> > * added `SIGINT` where there is `trap` usage
> > ```
> 
> You have both rebased and added the changes, making it very hard now to see what changed.
> Please do not mix both when pushing.

Would rebase first then add changes be easier to review?

EdDev commented:
> Would rebase first then add changes be easier to review?

The order does not matter, only that fact that two steps exist:
- rebase, push
- add changes, push
(or vice versa)

phoracek commented:
/approve
